torch/csrc/distributed/rpc/request_callback_impl.cpp

```diff
@@ -168,15 +169,70 @@ void RequestCallbackImpl::processRpc(
     case MessageType::PYTHON_CALL: {
       auto& upc = static_cast<UnpickledPythonCall&>(rpc);
       auto& pythonRpcHandler = PythonRpcHandler::getInstance();
-      std::shared_ptr<SerializedPyObj> serializedPyObj = nullptr;
-      {
-        py::gil_scoped_acquire acquire;
-        serializedPyObj =
-            std::make_shared<SerializedPyObj>(pythonRpcHandler.serialize(
-                pythonRpcHandler.runPythonUdf(std::move(upc).movePythonUdf())));
+      if (upc.isAsyncFunction()) {
+        std::shared_ptr<jit::PythonFutureWrapper> pyFuture;
+        {
+          py::gil_scoped_acquire acquire;
+          auto result =
+              pythonRpcHandler.runPythonUdf(std::move(upc).movePythonUdf());
+
+          if (pythonRpcHandler.isException(result) ) {
+            // Hit exception when running the user function.
+            // Not releasing GIL before serialize to avoid an additional
+            // context switch.
+            auto serializedPyObj = pythonRpcHandler.serialize(result);
+            py::gil_scoped_release release;
+            markComplete(
+                std::move(PythonResp(std::move(serializedPyObj))).toMessage());
+            return;
+          }
+
+          try {
+            pyFuture = result.cast<std::shared_ptr<jit::PythonFutureWrapper>>();
+          } catch (...) {
+            // user function return type cannot be cast to a Future type.
+            auto type = result.get_type();
+            auto errMsg = c10::str(
+                "Functions decorated with @rpc.async_function must return a "
+                "torch.futures.Future object, but got ",
+                type.attr("__module__").cast<std::string>(),
+                ".",
+                type.attr("__qualname__").cast<std::string>());
+
+            py::gil_scoped_release release;
+            responseFuture->markCompleted(
+                createExceptionResponse(errMsg, messageId));
+            return;
+          }
+        }
+        pyFuture->fut->addCallback([msgId = messageId,
+                                    responseFuture = responseFuture,
+                                    jitFuture = pyFuture->fut,
+                                    &pythonRpcHandler]() {
+          std::shared_ptr<SerializedPyObj> serializedPyObj;
+          {
+            py::gil_scoped_acquire acquire;
+            serializedPyObj =
+                std::make_shared<SerializedPyObj>(pythonRpcHandler.serialize(
+                    jit::toPyObject(jitFuture->value())));
+          }
+          auto m =
+              std::move(PythonResp(std::move(*serializedPyObj))).toMessage();
+          m.setId(msgId);
+          responseFuture->markCompleted(std::move(m));
+        });
+      } else {
+        std::shared_ptr<SerializedPyObj> serializedPyObj;
```

Comment:
What happens if a UDF returns a future that's set later, but is not decorated properly? (I guess this q is even valid today, when we don't have this decorator). Do we need error handling for this?

Reply:
In that case, the RPC would try to pickle the Future, which would hit the pickle error. Let me add a test for that.